xtf: init at 0-unstable-2024-07-25

SigmaSquadron opened the pull request:
## Description of changes

- Adds the Xen Test Framework, a comprehensive suite of tests that can be run in a dom0 to test some advanced functionality of the Xen Hypervisor. It's a python script that calls the `xl` binary to run the tests defined in the tests/ directory.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
- [x] Tested, as applicable:
  - No automated tests, for the same reason as every other Xen-related package.
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 335855"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AndersonTorres commented on pkgs/by-name/xt/xtf/package.nix:
```diff
@@ -0,0 +1,62 @@
+  postInstall =
+    # Copy autogenerated documentation to $doc.
+    ''
+      mkdir -p $doc/share/doc/xtf
```
Use the more general variables:

```suggestion
      mkdir -p ''${!outputDoc}/share/doc/xtf
```

AndersonTorres commented on pkgs/by-name/xt/xtf/package.nix:
```diff
@@ -0,0 +1,62 @@
+    # This is necessary because xtf-runner should be in the same directory as the tests/ directory.
+    + ''
+      wrapPythonProgramsIn "$out/share/xtf" "$out $pythonPath"
+      mkdir -p $out/bin
```
```suggestion
      mkdir -p ''${!outputBin}/bin
```

AndersonTorres commented on pkgs/by-name/xt/xtf/package.nix:
```diff
@@ -0,0 +1,62 @@
+
+  # The Makefile writes the finished objects to $DESTDIR$xtfdir.
+  installFlags = [
+    "DESTDIR=$(out)"
+    "xtfdir=/share/xtf"
```
Why not

```suggestion
    "xtfdir=$(out)/share/xtf"
```

?

SigmaSquadron replied:
Because it'd then build to `/tmp/<temporary build directory>/build/xtf-f37c4574/dist/nix/store/<hash>-xtf-0-unstable-2024-07-25/share/xtf`

SigmaSquadron replied:
Actually, looks like we need to do this anyway, otherwise `xl` gets confused.

AndersonTorres commented on pkgs/by-name/xt/xtf/package.nix:
```diff
@@ -0,0 +1,62 @@
+  ];
+
+  postInstall =
+    # Copy autogenerated documentation to $doc.
```
Looks redundant.
```suggestion
```

SigmaSquadron replied:
I personally like trivial comments to explain basic things to people who aren't as familiar with the build system, but sure.

SigmaSquadron commented:
Result of `nixpkgs-review pr 335855` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>xtf</li>
    <li>xtf.dev</li>
    <li>xtf.doc</li>
  </ul>
</details>

SigmaSquadron commented:
Result of `nixpkgs-review pr 335855` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>xtf</li>
    <li>xtf.dev</li>
    <li>xtf.doc</li>
  </ul>
</details>

SigmaSquadron commented:
Dropped aarch64 as it doesn't compile.

SigmaSquadron commented:
Result of `xtf --all --host` run on Xen 4.19:
<details>

<summary>72 tests succeeded, 23 tests skipped:</summary>

```
test-hvm32-cpuid-faulting                SKIP
test-hvm32pae-cpuid-faulting             SKIP
test-hvm32pse-cpuid-faulting             SKIP
test-hvm64-cpuid-faulting                SKIP
test-pv64-cpuid-faulting                 SUCCESS
test-hvm64-fpu-exception-emulation       SKIP
test-hvm32-invlpg~hap                    SUCCESS
test-hvm32-invlpg~shadow                 SUCCESS
test-hvm32pae-invlpg~hap                 SUCCESS
test-hvm32pae-invlpg~shadow              SUCCESS
test-hvm64-invlpg~hap                    SUCCESS
test-hvm64-invlpg~shadow                 SUCCESS
test-hvm64-lbr-tsx-vmentry               SUCCESS
test-hvm32-livepatch-priv-check          SUCCESS
test-hvm64-livepatch-priv-check          SUCCESS
test-pv64-livepatch-priv-check           SUCCESS
test-hvm32-lm-ts                         SUCCESS
test-hvm64-lm-ts                         SUCCESS
test-hvm32pae-memop-seg                  SUCCESS
test-hvm64-memop-seg                     SUCCESS
test-pv64-memop-seg                      SUCCESS
test-hvm32pae-nmi-taskswitch-priv        SUCCESS
test-pv64-pv-fsgsbase                    SUCCESS
test-pv64-pv-iopl~hypercall              SUCCESS
test-pv64-pv-iopl~vmassist               SUCCESS
test-hvm32-swint-emulation               SKIP
test-hvm32pae-swint-emulation            SKIP
test-hvm32pse-swint-emulation            SKIP
test-hvm64-swint-emulation               SKIP
test-hvm32-umip                          SKIP
test-hvm64-umip                          SKIP
test-hvm32-xsa-122                       SUCCESS
test-hvm32pae-xsa-122                    SUCCESS
test-hvm32pse-xsa-122                    SUCCESS
test-hvm64-xsa-122                       SUCCESS
test-pv64-xsa-122                        SUCCESS
test-hvm32-xsa-123                       SKIP
test-pv64-xsa-167                        SKIP
test-hvm64-xsa-168~shadow                SUCCESS
test-hvm64-xsa-170                       SKIP
test-hvm64-xsa-173~shadow                SUCCESS
test-pv64-xsa-182                        SUCCESS
test-hvm32-xsa-186                       SKIP
test-hvm64-xsa-186                       SKIP
test-hvm32-xsa-188                       SUCCESS
test-hvm32pae-xsa-188                    SUCCESS
test-hvm32pse-xsa-188                    SUCCESS
test-hvm64-xsa-188                       SUCCESS
test-pv64-xsa-188                        SUCCESS
test-hvm32-xsa-191                       SKIP
test-hvm32-xsa-192                       SUCCESS
test-pv64-xsa-193                        SUCCESS
test-hvm64-xsa-195                       SUCCESS
test-hvm64-xsa-196                       SKIP
test-hvm32-xsa-200                       SKIP
test-hvm32-xsa-203                       SKIP
test-hvm64-xsa-204                       SKIP
test-pv64-xsa-212                        SUCCESS
test-pv64-xsa-213                        SUCCESS
test-hvm64-xsa-221                       SUCCESS
test-pv64-xsa-221                        SUCCESS
test-pv64-xsa-224                        SUCCESS
test-pv64-xsa-227                        SUCCESS
test-hvm64-xsa-231                       SUCCESS
test-pv64-xsa-231                        SUCCESS
test-hvm64-xsa-232                       SUCCESS
test-pv64-xsa-232                        SUCCESS
test-pv64-xsa-234                        SUCCESS
test-hvm32-xsa-239                       SUCCESS
test-pv64-xsa-255                        SUCCESS
test-pv64-xsa-259                        SUCCESS
test-pv64-xsa-260                        SUCCESS
test-hvm64-xsa-261                       SUCCESS
test-pv64-xsa-265                        SUCCESS
test-hvm64-xsa-269                       SUCCESS
test-hvm64-xsa-277                       SUCCESS
test-hvm64-xsa-278                       SUCCESS
test-pv64-xsa-279                        SUCCESS
test-pv64-xsa-286                        SUCCESS
test-pv64-xsa-296                        SUCCESS
test-pv64-xsa-298                        SUCCESS
test-hvm64-xsa-304                       SUCCESS
test-hvm64-xsa-308                       SUCCESS
test-pv64-xsa-316                        SUCCESS
test-hvm32-xsa-317                       SUCCESS
test-hvm32pae-xsa-317                    SUCCESS
test-hvm32pse-xsa-317                    SUCCESS
test-hvm64-xsa-317                       SUCCESS
test-pv64-xsa-317                        SUCCESS
test-pv64-xsa-333                        SUCCESS
test-pv64-xsa-339                        SUCCESS
test-pv64-xsa-444                        SKIP
test-hvm64-xsa-451                       SKIP
test-hvm64-xsa-454                       SUCCESS
test-hvm32pae-xsa-consoleio-write        SUCCESS
```

</details>

SigmaSquadron commented:
- Removed `DESTDIR` override; we need `xtfdir` to include the `$out` path anyway.
- Added `graphviz` so we can generate dot graphs for the documentation.

SigmaSquadron commented:
Did someone break `master` again?

@ofborg eval

SigmaSquadron committed:
xtf: init at 0-unstable-2024-07-25

The Xen Test Framework, a comprehensive suite of tests that can be run
in a dom0 to test some advanced functionality of the Xen Hypervisor.

It's a python script that calls the `xl` binary to run the tests defined
in the tests/ directory.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: add xtf to README

Indicates that XTF should be used as a baseline test suite for any Xen
update from now on.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

AndersonTorres reviewed (approved):
Result of `nixpkgs-review pr 335855` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>xtf</li>
    <li>xtf.dev</li>
    <li>xtf.doc</li>
  </ul>
</details>
